src/plugins/auth/local.js

```diff
@@ -0,0 +1,42 @@
+"use strict";
+
+const Helper = require("../../helper");
+const colors = require("colors/safe");
+
+function localAuth(manager, client, user, password, callback) {
+	// If no user is found, or if the client has not provided a password,
+	// fail the authentication straight away
+	if (!client || !password) {
+		return callback(false);
+	}
+
+	// If this user has no password set, fail the authentication
+	if (!client.config.password) {
+		log.error(`User ${colors.bold(user)} with no local password set tried to sign in. (Probably a LDAP user)`);
+		return callback(false);
+	}
+
+	Helper.password
+		.compare(password, client.config.password)
+		.then((matching) => {
+			if (matching && Helper.password.requiresUpdate(client.config.password)) {
+				const hash = Helper.password.hash(password);
+
+				client.setPassword(hash, (success) => {
+					if (success) {
+						log.info(`User ${colors.bold(client.name)} logged in and their hashed password has been updated to match new security requirements`);
+					}
+				});
+			}
+
+			callback(matching);
+		}).catch((error) => {
+			log.error(`Error while checking users password. Error: ${error}`);
+		});
+}
+
+module.exports = {
+	auth: localAuth,
+	isEnabled: () => true
```

Comment:
Is it expected that local auth is enabled even when LDAP is enabled?

Reply:
Looking at code below, because it's in an `else if` that will be ignored when LDAP auth is enabled, this gets ignored. I would prefer if explicit:

```js
isEnabled: !Helper.config.public && !Helper.config.ldap.enable
```

(Assuming not being a function works, `() => ...` otherwise)

Reply:
This is intented to be a fallback mechanism. The LDAP auth plugin would be the first tested, then if it is not available the system falls back to the local one. There might be an arbitrary list of auth plugin to test, and the local auth plugin is intended to be the ultimate one, that should hence always be available. So, `isEnabled: () => true` sounds semantically more appropriate to me.

Reply:
I'll replace the if else by the mechanism I just described to make things clearer.

Reply:
@astorije See https://github.com/eliemichel/lounge/commit/dfa5bffff6c3dfb37cc0c8a7806e3a28c18e1463

Reply:
I get what you're coming from, and I agree with most of it, except with the idea of an ultimate default one. I think that if you set up The Lounge to use LDAP, and it actually uses local auth as a fallback, it could be bad.
If you want to use an auth mechanism and only this one, but it fails for some reason, then The Lounge should be failing instead. It's unlikely that you have set up local auth with correct users if you are intending to use LDAP anyway.

But we're thinking way ahead so I'm fine with the current state :)
There is no notion of "not available" for LDAP in this PR, I think, so it doesn't matter.